packages/rollup-contracts/contracts/state-tree/FullPatriciaTree.sol

```diff
@@ -0,0 +1,380 @@
+            // So, this sets the bitsTraversed'th bit in the branch mask to 1.
+            branchMask |= uint(1) << (255 - bitsTraversed);
+
+            // As explained in the last line, we traverse left or right based on the next bit.
+            uint head;
+            D.Label memory tail;
+            (head, tail) = Utils.chopFirstBit(suffix);
+            // head, either 0 or 1, tells us which child edge is to traverse to, and which is the sibling to supply in our proof.
+            uint siblingIndex = 1 - head;
+            siblings[numSiblings++] = edgeHash(
+                tree.nodes[currentEdge.node].children[siblingIndex]
+            );
+            // Now, update the current edge to be processed in next iteration
+            currentEdge = tree.nodes[currentEdge.node].children[head];
+            // Account for having processed another bit by choosing left or right.
+            bitsTraversed += 1;
+            remaining = tail;
+        }
+        if (numSiblings > 0)
+        {
+            _siblings = new bytes32[](numSiblings);
+            for (uint i = 0; i < numSiblings; i++)
+                _siblings[i] = siblings[i];
+        }
+    }
+
+    function getNonInclusionProof(Tree storage tree, bytes32 key) internal view returns (
+        D.Label memory potentialSiblingCumulativeLabel,
+        bytes32 potentialSiblingValue,
+        uint branchMask,
+        bytes32[] memory _siblings
+    ){
+        uint length;
+        uint numSiblings;
+
+        D.Label memory cumulativeKeyLabel = D.Label(key, 256);
+
+        // Start from root edge
+        D.Label memory label = cumulativeKeyLabel;
+        D.Edge memory e = tree.rootEdge;
+        bytes32[256] memory siblings;
+
+        while (true) {
+            // Find at edge
+            require(label.length >= e.label.length);
+            D.Label memory prefix;
+            D.Label memory suffix;
+            (prefix, suffix) = Utils.splitCommonPrefix(label, e.label);
+
+            // suffix.length == 0 means that the key exists. Thus the length of the suffix should be not zero
+            require(suffix.length != 0);
+
+            if (prefix.length >= e.label.length) {
+                // Partial matched, keep finding
+                length += prefix.length;
+                branchMask |= uint(1) << (255 - length);
+                length += 1;
+                uint head;
+                (head, label) = Utils.chopFirstBit(suffix);
+                siblings[numSiblings++] = edgeHash(tree.nodes[e.node].children[1 - head]);
+                e = tree.nodes[e.node].children[head];
+            } else {
+                // Found the potential sibling. Set data to return
+                (D.Label memory sharedCumulativePrefix, ) = Utils.splitAt(cumulativeKeyLabel, length);
+                potentialSiblingCumulativeLabel = Utils.combineLabels(sharedCumulativePrefix, e.label);
+                potentialSiblingValue = e.node;
```

Comment:
I'm confused by _potential_ as well as the fact that `siblings[numSiblings]` should have already been set  to be the proper sibling of `e` last iteration of the loop, but now we're calling `e` the potential sibling? Shouldn't `e` be an ancestor of the non-existent key we're looking for?